src/apps/vpn/platforms/linux/daemon/dbusservice.cpp

```diff
@@ -258,8 +291,126 @@ bool DBusService::firewallPid(int rootpid, const QString& state) {
+  // of the network interfaces and Wireguard tunnel.
+  //
+  // Acquire CAP_SETUID, we need it to masquerade as other users on their
+  // session busses for application tracking.
+  //
+  // NOTE: ptrace is a dangerous permission to hold. If it may be safer to
+  // relent on the executable check and grant CAP_NET_ADMIN to the client
+  // process during installation.
+  //
+  // Clear all other capabilities, effectively discarding our root permissions.
+  cap_value_t newcaps[] = {CAP_NET_ADMIN, CAP_SETUID};
+  int numcaps = sizeof(newcaps) / sizeof(cap_value_t);
+  if (cap_set_flag(caps, CAP_EFFECTIVE, numcaps, newcaps, CAP_SET) ||
+      cap_set_flag(caps, CAP_PERMITTED, numcaps, newcaps, CAP_SET)) {
+    logger.warning() << "Failed to set process capability flags";
+    return;
+  }
+  if (cap_set_proc(caps) != 0) {
+    logger.warning() << "Failed to update process capabilities";
+    return;
+  }
+}
+
+/* Checks to see if the caller has sufficient authorization */
+bool DBusService::checkCallerAuthz() {
+  if (!calledFromDBus()) {
+    // If this is not a D-Bus call, it came from the daemon itself.
+    return true;
+  }
+  QDBusConnectionInterface* iface = QDBusConnection::systemBus().interface();
+
+  // If the VPN is active, and we know the UID that turned it on, as a special
+  // case we permit that user full access to the D-Bus API in order to manage
+  // the connection.
+  if (m_sessionUid != 0) {
+    QDBusReply<uint> reply = iface->serviceUid(message().service());
+    if (reply.isValid() && m_sessionUid == reply.value()) {
+      return true;
+    }
+  }
+  // Otherwise, if this is the activate method, we permit any non-root user to
+  // activate the VPN, but we will remember their UID for later authentication.
+  else if ((message().type() == QDBusMessage::MethodCallMessage) &&
+           (message().member() == "activate")) {
+    QDBusReply<uint> reply = iface->serviceUid(message().service());
+    uint senderuid = reply.value();
+    if (reply.isValid() && senderuid != 0) {
+      m_sessionUid = senderuid;
+      return true;
+    }
+  }
+  // In all other cases, the use of this D-Bus API requires the CAP_NET_ADMIN
+  // permission, which we can check by examining the PID of the sender. Note
+  // that a zero UID (root) is used as a guard value to fall back to this case.
+
+  // Get the PID of the D-Bus message sender.
+  QDBusReply<uint> reply = iface->servicePid(message().service());
+  uint senderpid = reply.value();
+  if (!reply.isValid() || (senderpid == 0)) {
+    // Could not lookup the sender's PID. Rejected!
+    logger.warning() << "Failed to resolve sender PID";
+    return false;
+  }
+
+  // Get the capabilties of the sender process.
+  cap_t caps = cap_get_pid(senderpid);
+  if (caps == nullptr) {
+    logger.warning() << "Failed to retrieve process capabilities";
+    return false;
+  }
+  auto guard = qScopeGuard([&] { cap_free(caps); });
```

Comment:
If the lambda expression isn't being called further down, can we remove the `guard` assignment?

Reply:
The scope guard is used to ensure that `cap_free()` is called at function exit, regardless of which path we take to exit the function.